src/stan/mcmc/chainset.hpp

```diff
@@ -0,0 +1,556 @@
+#ifndef STAN_MCMC_CHAINSET_HPP
+#define STAN_MCMC_CHAINSET_HPP
+
+#include <stan/io/stan_csv_reader.hpp>
+#include <stan/math/prim.hpp>
+#include <stan/analyze/mcmc/split_rank_normalized_ess.hpp>
+#include <stan/analyze/mcmc/split_rank_normalized_rhat.hpp>
+#include <boost/accumulators/accumulators.hpp>
+#include <boost/accumulators/statistics/stats.hpp>
+#include <boost/accumulators/statistics/mean.hpp>
+#include <boost/accumulators/statistics/tail_quantile.hpp>
+#include <boost/accumulators/statistics/p_square_quantile.hpp>
+#include <boost/accumulators/statistics/variance.hpp>
+#include <boost/accumulators/statistics/covariance.hpp>
+#include <boost/accumulators/statistics/variates/covariate.hpp>
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <sstream>
+#include <utility>
+#include <vector>
+#include <cstdlib>
+
+namespace stan {
+namespace mcmc {
+using Eigen::Dynamic;
+
+/**
+ * An <code>mcmc::chainset</code> object manages the post-warmup draws
+ * across a set of MCMC chains, which all have the same number or samples.
+ *
+ * <p><b>Storage Order</b>: Storage is column/last-index major.
+ */
+class chainset {
+ private:
+  size_t num_samples_;
+  std::vector<std::string> param_names_;
+  std::vector<Eigen::MatrixXd> chains_;
```

Comment:
Instead of an std::vector of chains with inner matrices of size `(Samples, Parameters)` can we just have one large matrix of size `(Samples * Chains, Parameters)`. I think this will be nicer for extracting information for each parameter

Reply:
this data structure is convenient for computing Rhat, because you need both w/in chain and cross-chain variances.